qiskit/backends/local/statevector_simulator_cpp.py

```diff
@@ -0,0 +1,93 @@
+from ._simulatorerror import SimulatorError
+
+logger = logging.getLogger(__name__)
+
+
+class StatevectorSimulatorCpp(QasmSimulatorCpp):
+    """C++ statevector simulator"""
+
+    DEFAULT_CONFIGURATION = {
+        'name': 'local_statevector_simulator_cpp',
+        'url': 'https://github.com/QISKit/qiskit-sdk-py/src/qasm-simulator-cpp',
+        'simulator': True,
+        'local': True,
+        'description': 'A C++ statevector simulator for qobj files',
+        'coupling_map': 'all-to-all',
+        'basis_gates': 'u1,u2,u3,cx,cz,id,x,y,z,h,s,sdg,t,tdg,rzz,snapshot'
+    }
+
+    def __init__(self, configuration=None):
+        super().__init__(configuration or self.DEFAULT_CONFIGURATION.copy())
+
+    def run(self, q_job):
+        """Run a QuantumJob on the backend."""
+        qobj = q_job.qobj
+        self._validate(qobj)
+        final_state_key = 32767  # Internal key for final state snapshot
+        # Add final snapshots to circuits
+        for circuit in qobj['circuits']:
+            circuit['compiled_circuit']['operations'].append(
+                {'name': 'snapshot', 'params': [final_state_key]})
+        result = super().run(q_job)._result
+        # Extract final state snapshot and move to 'quantum_state' data field
+        for res in result['result']:
+            snapshots = res['data']['snapshots']
+            if str(final_state_key) in snapshots:
+                final_state_key = str(final_state_key)
+            # Pop off final snapshot added above
+            final_state = snapshots.pop(final_state_key, None)
+            final_state = final_state['quantum_state'][0]
+            # Add final state to results data
+            res['data']['quantum_state'] = final_state
+            # Remove snapshot dict if empty
+            if snapshots == {}:
+                res['data'].pop('snapshots', None)
+        return Result(result, qobj)
+
+    def _validate(self, qobj):
+        """Semantic validations of the qobj which cannot be done via schemas.
+        Some of these may later move to backend schemas.
+
+        1. No shots
+        2. No measurements in the middle
+        """
+        if qobj['config']['shots'] != 1:
+            logger.warning("statevector simulator only supports 1 shot. "
+                           "Setting shots=1.")
+            qobj['config']['shots'] = 1
+        for circuit in qobj['circuits']:
+            if 'shots' in circuit['config'] and circuit['config']['shots'] != 1:
+                logger.warning("statevector simulator only supports 1 shot. "
+                               "Setting shots=1 for circuit %s", circuit['name'])
+                circuit['config']['shots'] = 1
+            for op in circuit['compiled_circuit']['operations']:
+                if op['name'] == 'measure':
+                    raise SimulatorError("In circuit {}: statevector simulator does "
+                                         "not support measure.".format(circuit['name']))
+        return
```

Comment:
This validation should also raise an error if there are `reset` operations int he circuit.